Apply workflow-enforcing code as local pre-push hook

creideiki opened the pull request:
Refactor the main code as a library function, and add thin wrappers to adapt to the different hook API:s (`pre-receive` for previous server-side mode, `pre-push` for new client-side mode.

creideiki committed:
Apply workflow-enforcing code as local pre-push hook

Refactor the main code as a library function, and add thin wrappers to
adapt to the different hook API:s (pre-receive for previous server-
side mode, pre-push for new client-side mode.

notCalle commented on git-hooks/enforce-workflow/pre-push:
```diff
@@ -0,0 +1,26 @@
+
+require_relative 'enforce-workflow.rb'
+
+puts '*** Running local pre-push tests...'
```
Drop the debug spam.

notCalle commented on git-hooks/enforce-workflow/pre-push:
```diff
@@ -0,0 +1,26 @@
+  enforce_workflow(rev_old, rev_new, ref)
+end
+
+puts '*** Local pre-push tests complete, pushing...'
```
Drop the debug spam.

notCalle commented on git-hooks/enforce-workflow/pre-receive:
```diff
@@ -0,0 +1,28 @@
+
+require_relative 'enforce-workflow.rb'
+
+puts '*** Running server-side pre-receive tests...'
```
Drop the debug spam.

notCalle commented on git-hooks/enforce-workflow/pre-receive:
```diff
@@ -0,0 +1,28 @@
+  enforce_workflow(rev_old, rev_new, ref)
+end
+
+puts '*** Server-side pre-receive tests complete, updating references...'
```
Drop the debug spam.

notCalle reviewed (changes requested):
Git already prefixes remote hook output with `remote: `, so no need for extra spam to tell local from remote hook output.

notCalle commented:
For the `pre-push` form, when you don't have access to the collaborators.yaml, stored on the remote side, instead of giving an unfriendly stack trace:
 - autogenerate authorization list from previously accepted commit signatures?
 - ignore gpg signature authorization check (only validate existence)?
 - add options to enable one or both of the above?

creideiki committed:
Remove messages saying which tests are local vs. remote

creideiki committed:
Wrap the workflow enforcer in a class

We want to be able to globally change the behaviour of the workflow
enforcer. Turning it into a class means we can use class variables to
do that later.

creideiki committed:
Add ignore_collaborators class variable

Even if you don't have the collaborators.yaml file available, it would
still be useful to be able to run the rest of the tests. To this end,
add a class variable "ignore_collaborators" which, when set, skips
the loading of the collaborators.yaml file and accepts all signers.
The GnuPG signatures still need to validate.

creideiki committed:
Ignore collaborators check in pre-push mode

The collaborators.yaml file is server-side configuration, so it's not
guaranteed to be available to clients. Therefore, run the pre-push
hook with ignore_collaborators set to true, to assume that all
signers would be accepted.

notCalle reviewed (changes requested):
I dislike this use of a class variable, when an instance variable would suffice, and rubocop agrees.

creideiki committed:
Add README file

Move existing documentation from comments at the top of
workflow_enforcer.rb to README.

notCalle reviewed (commented):
I can see use cases when a populated `collaborators.yaml` is useful in the `pre-push` hook, e.g. for GitHub work, where server-side hooks are not available. Maybe change to `ignore_missing_collaborators` instead?

creideiki committed:
Make options instance variables

Instead of having behaviour-modifying options as class variables, make
them instance variables, set by passing options to initialize.

creideiki committed:
Add ignore_missing_collaborators mode, default for pre-push

Add an option to still load collaborators.yaml, but let failures of
the check be just warnings instead of fatal errors. Make this mode
default for the pre-push hook.

creideiki committed:
Take option arguments as kwargs hash

To avoid having to specify the keyword arguments yet again, take them
as an implicit kwargs hash.

creideiki committed:
Add ignore_missing_collaborators_file mode, default for pre-push

For workflows where you can't add hooks to the server, e.g. GitHub,
allow using the client-side hook in an enforcing mode if the
collaborators.yaml file is present locally.
